source/_posts/2015-11-16-immutable-strings-in-ruby-2-dot-3.markdown

```diff
@@ -0,0 +1,279 @@
+
+This approach will help to avoid unnecessary data duplication, therefore
+unnecessary object allocation is prevented and it will imply that the program
+runs faster because it will spend less time in the Garbage Collector.
+
+Furthermore, when we're concurrently accessing a shared literal across several
+threads, there is no need to synchronize it access. This could lead to simpler
+multi-threaded programs.
+
+# Moving towards immutability
+
+On this initial phase of Ruby's strings immutability, we can enable this feature
+at a project level through a flag or on a per file basis using a pragma. Furthermore
+for a more detailed error stack (while debugging) another flag can be enabled.
+
+If you're working on an existent project and want to introduce this new feature,
+obviously the most appropriate method is to use the pragma approach. Just add the
+following comment at the beginning of each file you want to switch to immutable
+strings and your done. Well, not really. Probably you will have some errors to fix.
+
+    # frozen_string_literal: true
+
+But, if you woke up today feeling like a kamikaze, enable frozen strings for the
+whole project and prepare yourself for some serious bug hunting. So, to enable
+string immutability for a project add the following flag when running your app.
+
+    --enable-frozen-string-literal
+
+One way or another, it's quite probable that you will encounter some bugs.
+Mainly because you are relying in 3rd party gems that are not prepared for
+immutable strings. Ruby ecosystem will need some time to adapt to this new
+feature, once most popular gems start to adapt frozen string literals the
+transition will be smoother.
+
+Let's run a quick example, create a file named `replace.rb` with the following
+lines:
+
+{% codeblock lang:ruby %}
+  str = "abcde"
+  str.gsub!("c", "b")
+{% endcodeblock %}
+
+Running it with frozen string literal flag enabled, will throw an error like the one below.
+
+    ruby --enable-frozen-string-literal replace.rb
+
+    replace.rb:2:in `gsub!': can't modify frozen String (RuntimeError)
+            from replace.rb:2:in `<main>'
+
+Finding were the frozen string was created on a small example is straightforward, but what if we're on a large
+codebase trying to modify a string that has been passed around between several methods distributed
+on multiple files? Finding the original string can be cumbersome. That's why a debug
+flag was added, add the debug flag as shown below and let's try again.
+
+    ruby --enable-frozen-string-literal --enable-frozen-string-literal-debug replace.rb
+
+    replace.rb:2:in `gsub!': can't modify frozen String, created at replace.rb:1 (RuntimeError)
+      from replace.rb:2:in `<main>'
+
+As you can see a subtle change is made to the error output, now we have exact
+information about were the string was created.
+
+### Same value literals points to the same object
+
+Same value literals points to the same object, i.e. `a` and `a` points to the
```

Comment:
Not sure how this ` tick renders in the blog anyway why not using single or double quotes which is the way the string is written in Ruby.